### ISSUE TITLE
Make ? work without :

### ISSUE THREAD

**Issue by davidanthoff** [JuliaLang/julia#6823]
There was a discussion about this pattern in Julia on the mailing list recently:

```
...
flag = true
...
flag && (x = rand(2,2))
...
```

I very much see the need for something short and sweet in these cases (rather than a long if ... etc), but this particular syntax strikes me as pretty opaque to new users. I've also seen things like

```
...
flag = true
...
flag && error("Error")
...
```

Which again is super useful for e.g. argument error checking, but somehow to me looks like a misuse of &&.

Couldn't ? without a : work in these cases? I.e. the following syntax:

```
...
flag = true
...
flag ? x = rand(2,2)
...
```

and

```
flag = true
...
flag ? error("error")
...
```

Essentially ? would just be a shorthand notation for an if.

**Comment by davidanthoff** [comment]
Is there some new issue that tracks suggestions like ``if a then b``, or some other alternative? I didn't get the impression that there was overwhelming dislike for some solution to this in the discussion here, so not clear to me why this issue is being closed.

**Comment by p-i-** [comment]
If we drop the leading `'if`, we would be able to do:
```julia
a then b
a else b
```
... both of which I think people will find perfectly clear / readable / obvious without explanation.

Thread here: https://discourse.julialang.org/t/proposal-then-else-syntax-to-replace-short-circuiting

**Comment by JeffBezanson** [comment]
I don't think that `else` syntax is workable, since the following currently parses:

```
if 1 < 0
2 else 3
end
```

or more generally when writing conditional expressions on a single line, `if 1 < 0 2 else 3 end`.

**Comment by StefanKarpinski** [comment]
@p-i-: that's basically already how I read `&&` and `||` in my head. To me the only real issue with `&&` and `||` is precedence issues with assignment (i.e. `x = y && z` vs `x && y = z`).

**Comment by StefanKarpinski** [comment]
Looking at the [precedence table](https://github.com/JuliaLang/julia/blob/cfac61d88625e3d58b9f0dbd4d9959e640b2a9f0/src/julia-parser.scm#L9-L17), I'm starting to think the Perl solution of having two different pairs of short-circuit operators for this is actually not so bad: `&&` and `||` have higher precedence than arrows, conditionals or assignment (as they currently do) while `and` and `or` have lower precedence than assignment. However, we have the added complication of `xor`, `&` and `|` (as well as `.&` and `.|`) so there's a lot of things to sort out. Basically, I think what we have now is not so bad.

**Comment by jballanc** [comment]
Perhaps consider borrowing Ruby's post-fix conditionals? `b if a` and `b unless a`? I know opinions on this particular feature vary, but I've always found them extremely useful.

**Comment by andyferris** [comment]
Sorry to revisit this, but I was thinking about how some of this stuff is parsed.

In some contexts at least, I can make the Julia code

```julia
 ;
```
which returns `nothing`. Given that lack of code is basically `nothing`, I'm wondering if it would be natural that I could write some code like:
```julia
cond ? a : ;
cond ? : b
```

(Perhaps we could have shorter forms like `cond ? a;` and `cond ?: b`... these aren't really a very readable alternative to #16389 but do seem to extend the ternary to binary forms in a _somewhat_ consistent way)

Current behavior (v0.6 and v0.7):
```julia
julia> true ? 1 : ;
ERROR: syntax: unexpected ";"

julia> true ? : 1
ERROR: syntax: space not allowed after ":" used for quoting
```

**Comment by rikhuijzer** [comment]
I would politely like to bump this discussion again since I think the use of 
```
A || do_stuff() 
```
meaning
```
if !(A); do_stuff(); end
``` 
and 
```
B && do_stuff()
``` 
meaning
```
if B; do_stuff(); end
```
is just super confusing. I would argue that users **expect** to find whether the clause is negated **in front** of the clause. With this style, readers have to infer the negation from **behind** the clause. Also, normally the conditional has no side effects and this promotes having side effects in conditionals. As an extreme example, this approach would also suggest to new users that 
```
if A
    B(x)
    C(y)
end
```
with
```
B(x)::Bool = x < 3
C(x)::Bool = 9 < y
```
should be rewritten to 
```
A && B(x) && C(y)
```
since it requires fewer lines of code.

**Comment by adienes** [comment]
closing for same reason as https://github.com/JuliaLang/julia/issues/16389#issuecomment-3393465739
